In [1]:
# import libraries
import pandas as pd
import numpy as np
import seaborn as sns

from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report, roc_auc_score

In [2]:
df = pd.read_csv('data.csv')
df

Age Attrition     BusinessTravel  DailyRate              Department  \
0      41       Yes      Travel_Rarely       1102                   Sales   
1      49        No  Travel_Frequently        279  Research & Development   
2      37       Yes      Travel_Rarely       1373  Research & Development   
3      33        No  Travel_Frequently       1392  Research & Development   
4      27        No      Travel_Rarely        591  Research & Development   
...   ...       ...                ...        ...                     ...   
1465   36        No  Travel_Frequently        884  Research & Development   
1466   39        No      Travel_Rarely        613  Research & Development   
1467   27        No      Travel_Rarely        155  Research & Development   
1468   49        No  Travel_Frequently       1023                   Sales   
1469   34        No      Travel_Rarely        628  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  \
0                    1          2  Life Sciences              1   
1                    8          1  Life Sciences              1   
2                    2          2          Other              1   
3                    3          4  Life Sciences              1   
4                    2          1        Medical              1   
...                ...        ...            ...            ...   
1465                23          2        Medical              1   
1466                 6          1        Medical              1   
1467                 4          3  Life Sciences              1   
1468                 2          3        Medical              1   
1469                 8          3        Medical              1   

      EmployeeNumber  ...  RelationshipSatisfaction StandardHours  \
0                  1  ...                         1            80   
1                  2  ...                         4            80   
2                  4  ...                         2            80   
3                  5  ...                         3            80   
4                  7  ...                         4            80   
...              ...  ...                       ...           ...   
1465            2061  ...                         3            80   
1466            2062  ...                         1            80   
1467            2064  ...                         2            80   
1468            2065  ...                         4            80   
1469            2068  ...                         1            80   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                    0                  8                      0   
1                    1                 10                      3   
2                    0                  7                      3   
3                    0                  8                      3   
4                    1                  6                      3   
...                ...                ...                    ...   
1465                 1                 17                      3   
1466                 1                  9                      5   
1467                 1                  6                      0   
1468                 0                 17                      3   
1469                 0                  6                      3   

     WorkLifeBalance  YearsAtCompany YearsInCurrentRole  \
0                  1               6                  4   
1                  3              10                  7   
2                  3               0                  0   
3                  3               8                  7   
4                  3               2                  2   
...              ...             ...                ...   
1465               3               5                  2   
1466               3               7                  7   
1467               3               6                  2   
1468               2               9    

In [3]:
df.describe()

Age    DailyRate  DistanceFromHome    Education  EmployeeCount  \
count  1470.000000  1470.000000       1470.000000  1470.000000         1470.0   
mean     36.923810   802.485714          9.192517     2.912925            1.0   
std       9.135373   403.509100          8.106864     1.024165            0.0   
min      18.000000   102.000000          1.000000     1.000000            1.0   
25%      30.000000   465.000000          2.000000     2.000000            1.0   
50%      36.000000   802.000000          7.000000     3.000000            1.0   
75%      43.000000  1157.000000         14.000000     4.000000            1.0   
max      60.000000  1499.000000         29.000000     5.000000            1.0   

       EmployeeNumber  EnvironmentSatisfaction   HourlyRate  JobInvolvement  \
count     1470.000000              1470.000000  1470.000000     1470.000000   
mean      1024.865306                 2.721769    65.891156        2.729932   
std        602.024335                 1.093082    20.329428        0.711561   
min          1.000000                 1.000000    30.000000        1.000000   
25%        491.250000                 2.000000    48.000000        2.000000   
50%       1020.500000                 3.000000    66.000000        3.000000   
75%       1555.750000                 4.000000    83.750000        3.000000   
max       2068.000000                 4.000000   100.000000        4.000000   

          JobLevel  ...  RelationshipSatisfaction  StandardHours  \
count  1470.000000  ...               1470.000000         1470.0   
mean      2.063946  ...                  2.712245           80.0   
std       1.106940  ...                  1.081209            0.0   
min       1.000000  ...                  1.000000           80.0   
25%       1.000000  ...                  2.000000           80.0   
50%       2.000000  ...                  3.000000           80.0   
75%       3.000000  ...                  4.000000           80.0   
max       5.000000  ...                  4.000000           80.0   

       StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
count       1470.000000        1470.000000            1470.000000   
mean           0.793878          11.279592               2.799320   
std            0.852077           7.780782               1.289271   
min            0.000000           0.000000               0.000000   
25%            0.000000           6.000000               2.000000   
50%            1.000000          10.000000               3.000000   
75%            1.000000          15.000000               3.000000   
max            3.000000          40.000000               6.000000   

       WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
count      1470.000000     1470.000000         1470.000000   
mean          2.761224        7.008163            4.229252   
std           0.706476        6.126525            3.623137   
min           1.000000        0.000000            0.000000   
25%           2.000000        3.000000            2.000000   
50%           3.000000        5.000000            3.000000   
75%           3.000000        9.000000            7.000000   
max           4.000000       40.000000           18.000000   

       YearsSinceLastPromotion  YearsWithCurrManager  
count              1470.000000           1470.000000  
mean                  2.187755              4.123129  
std                   3.222430              3.568136  
min                   0.000000              0.000000  
25%                   0.000000              2.000000  
50%                   1.000000              3.000000  
75%                   3.000000              7.000000  
max                  15.000000             17.000000  

[8 rows x 26 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

In [5]:
df.describe(include=object).T

count unique                     top  freq
Attrition       1470      2                      No  1233
BusinessTravel  1470      3           Travel_Rarely  1043
Department      1470      3  Research & Development   961
EducationField  1470      6           Life Sciences   606
Gender          1470      2                    Male   882
JobRole         1470      9         Sales Executive   326
MaritalStatus   1470      3                 Married   673
Over18          1470      1                       Y  1470
OverTime        1470      2                      No  1054

In [6]:
df.groupby('Attrition').size()

Attrition
No     1233
Yes     237
dtype: int64

In [7]:
df['Attrition'] = df['Attrition'].factorize(['No','Yes'])[0]
df.head()

Age  Attrition     BusinessTravel  DailyRate              Department  \
0   41          1      Travel_Rarely       1102                   Sales   
1   49          0  Travel_Frequently        279  Research & Development   
2   37          1      Travel_Rarely       1373  Research & Development   
3   33          0  Travel_Frequently       1392  Research & Development   
4   27          0      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0  ...                         1            80                 0   
1  ...                         4            80                 1   
2  ...                         2            80                 0   
3  ...                         3            80                 0   
4  ...                         4            80                 1   

   TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                  8                      0               1               6   
1                 10                      3               3              10   
2                  7                      3               3               0   
3                  8                      3               3               8   
4                  6                      3               3               2   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                  4                        0                     5  
1                  7                        1                     7  
2                  0                        0                     0  
3                  7                        3                     0  
4                  2                        2                     2  

[5 rows x 35 columns]

In [8]:
for column in df.columns:
    print(f"{column}: Number of unique values {df[column].nunique()}")
    print("==========================================================")

Age: Number of unique values 43
Attrition: Number of unique values 2
BusinessTravel: Number of unique values 3
DailyRate: Number of unique values 886
Department: Number of unique values 3
DistanceFromHome: Number of unique values 29
Education: Number of unique values 5
EducationField: Number of unique values 6
EmployeeCount: Number of unique values 1
EmployeeNumber: Number of unique values 1470
EnvironmentSatisfaction: Number of unique values 4
Gender: Number of unique values 2
HourlyRate: Number of unique values 71
JobInvolvement: Number of unique values 4
JobLevel: Number of unique values 5
JobRole: Number of unique values 9
JobSatisfaction: Number of unique values 4
MaritalStatus: Number of unique values 3
MonthlyIncome: Number of unique values 1349
MonthlyRate: Number of unique values 1427
NumCompaniesWorked: Number of unique values 10
Over18: Number of unique values 1
OverTime: Number of unique values 2
PercentSalaryHike: Number of unique values 15
PerformanceRating: Number of uni

We notice that 'EmployeeCount', 'Over18' and 'StandardHours' have only one unique values and 'EmployeeNumber' has 1470 unique values. This features aren't useful for us, So we are going to drop those columns.

In [9]:
#df.Attrition.replace(to_replace = dict(Yes = 1, No = 0), inplace = True)
df.drop(['EmployeeCount', 'EmployeeNumber', 'Over18', 'StandardHours'], axis="columns", inplace=True)
df.head()

Age  Attrition     BusinessTravel  DailyRate              Department  \
0   41          1      Travel_Rarely       1102                   Sales   
1   49          0  Travel_Frequently        279  Research & Development   
2   37          1      Travel_Rarely       1373  Research & Development   
3   33          0  Travel_Frequently       1392  Research & Development   
4   27          0      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EnvironmentSatisfaction  \
0                 1          2  Life Sciences                        2   
1                 8          1  Life Sciences                        3   
2                 2          2          Other                        4   
3                 3          4  Life Sciences                        4   
4                 2          1        Medical                        1   

   Gender  ...  PerformanceRating  RelationshipSatisfaction  StockOptionLevel  \
0  Female  ...                  3                         1                 0   
1    Male  ...                  4                         4                 1   
2    Male  ...                  3                         2                 0   
3  Female  ...                  3                         3                 0   
4    Male  ...                  3                         4                 1   

  TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                 8                      0               1               6   
1                10                      3               3              10   
2                 7                      3               3               0   
3                 8                      3               3               8   
4                 6                      3               3               2   

   YearsInCurrentRole  YearsSinceLastPromotion YearsWithCurrManager  
0                   4                        0                    5  
1                   7                        1                    7  
2                   0                        0                    0  
3                   7                        3                    0  
4                   2                        2                    2  

[5 rows x 31 columns]

In [15]:
# Transform categorical data into dummies
dummy_col = [column for column in df.drop('Attrition', axis=1).columns if df[column].nunique() < 20]
data = pd.get_dummies(df, columns=dummy_col, drop_first=True, dtype='uint8')
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Columns: 136 entries, Age to YearsWithCurrManager_17
dtypes: int64(9), uint8(127)
memory usage: 285.8 KB


In [16]:
X = data.drop('Attrition', axis=1)
y = data['Attrition']

In [17]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42, stratify=y)

In [18]:
scaler = StandardScaler()
X_train = pd.DataFrame(scaler.fit_transform(X_train), columns = X_train.columns)
X_test= pd.DataFrame(scaler.transform(X_test), columns = X_test.columns)
X_train.head()

Age  DailyRate  DistanceFromHome  HourlyRate  MonthlyIncome  \
0 -1.958919  -0.413973         -0.036578   -0.547908      -0.867017   
1 -0.430695   1.176070          0.699024   -0.449557       1.532088   
2  1.643324   1.296833          0.208623    0.829007      -0.737494   
3 -1.085648  -0.549831         -0.894779    1.074884      -0.860942   
4 -0.321536  -1.274408         -0.894779    0.484778      -0.936660   

   MonthlyRate  TotalWorkingYears  YearsAtCompany  \
0    -1.009049          -1.343310       -0.998443   
1     1.404772           0.478372        0.010293   
2     1.017664          -0.302349       -0.325952   
3    -1.130107          -0.692709       -0.494075   
4    -0.435997          -0.692709       -0.494075   

   BusinessTravel_Travel_Frequently  BusinessTravel_Travel_Rarely  ...  \
0                         -0.468182                      0.636975  ...   
1                         -0.468182                      0.636975  ...   
2                         -0.468182                      0.636975  ...   
3                          2.135922                     -1.569919  ...   
4                          2.135922                     -1.569919  ...   

   YearsWithCurrManager_8  YearsWithCurrManager_9  YearsWithCurrManager_10  \
0               -0.292306               -0.208831                -0.140789   
1               -0.292306               -0.208831                -0.140789   
2               -0.292306               -0.208831                -0.140789   
3               -0.292306               -0.208831                -0.140789   
4               -0.292306               -0.208831                -0.140789   

   YearsWithCurrManager_11  YearsWithCurrManager_12  YearsWithCurrManager_13  \
0                -0.125677                -0.121626                -0.088518   
1                -0.125677                -0.121626                -0.088518   
2                -0.125677                -0.121626                -0.088518   
3                -0.125677                -0.121626                -0.088518   
4                -0.125677                -0.121626                -0.088518   

   YearsWithCurrManager_14  YearsWithCurrManager_15  YearsWithCurrManager_16  \
0                 -0.06247                 -0.04413                -0.031189   
1                 -0.06247                 -0.04413                -0.031189   
2                 -0.06247                 -0.04413                -0.031189   
3                 -0.06247                 -0.04413                -0.031189   
4                 -0.06247                 -0.04413                -0.031189   

   YearsWithCurrManager_17  
0                 -0.06247  
1                 -0.06247  
2                 -0.06247  
3                 -0.06247  
4                 -0.06247  

[5 rows x 135 columns]

In [19]:
# Logistic Regression
lr_clf = LogisticRegression(solver='liblinear', penalty='l1')
lr_clf.fit(X_train, y_train)

print('TRAINING RESULTS')
print('=======================')
print('Classification report')
print(classification_report(y_train,lr_clf.predict(X_train)))
print('Accuracy score:',accuracy_score(y_train, lr_clf.predict(X_train)))
print('ROC AUC: ', roc_auc_score(y_train,lr_clf.predict(X_train)))
print('\nTESTING RESULTS')
print('=======================')
print('Classification report')
print(classification_report(y_test,lr_clf.predict(X_test)))
print('Accuracy score:',accuracy_score(y_test, lr_clf.predict(X_test)))
print('ROC AUC: ', roc_auc_score(y_test,lr_clf.predict(X_test)))

TRAINING RESULTS
Classification report
              precision    recall  f1-score   support

           0       0.94      0.98      0.96       863
           1       0.88      0.64      0.75       166

    accuracy                           0.93      1029
   macro avg       0.91      0.81      0.85      1029
weighted avg       0.93      0.93      0.92      1029

Accuracy score: 0.9290573372206026
ROC AUC:  0.8141779167655558

TESTING RESULTS
Classification report
              precision    recall  f1-score   support

           0       0.89      0.94      0.91       370
           1       0.56      0.39      0.46        71

    accuracy                           0.85       441
   macro avg       0.73      0.67      0.69       441
weighted avg       0.84      0.85      0.84       441

Accuracy score: 0.8526077097505669
ROC AUC:  0.6674533688618195


Logistic regression model have high accuracy score on training dataset and also high accuracy score on test dataset. Tested results with both small and large train dataset.

In [20]:
# Random Forest
rf_clf = RandomForestClassifier(n_estimators=50, bootstrap=False)
rf_clf.fit(X_train, y_train)

print('TRAINING RESULTS')
print('=======================')
print('Classification report')
print(classification_report(y_train,rf_clf.predict(X_train)))
print('Accuracy score:',accuracy_score(y_train, rf_clf.predict(X_train)))
print('ROC AUC: ', roc_auc_score(y_train,rf_clf.predict(X_train)))
print('\nTESTING RESULTS')
print('=======================')
print('Classification report')
print(classification_report(y_test,rf_clf.predict(X_test)))
print('Accuracy score:',accuracy_score(y_test, rf_clf.predict(X_test)))
print('ROC AUC: ', roc_auc_score(y_test,rf_clf.predict(X_test)))

TRAINING RESULTS
Classification report
              precision    recall  f1-score   support

           0       1.00      1.00      1.00       863
           1       1.00      1.00      1.00       166

    accuracy                           1.00      1029
   macro avg       1.00      1.00      1.00      1029
weighted avg       1.00      1.00      1.00      1029

Accuracy score: 1.0
ROC AUC:  1.0

TESTING RESULTS
Classification report
              precision    recall  f1-score   support

           0       0.85      0.98      0.91       370
           1       0.50      0.13      0.20        71

    accuracy                           0.84       441
   macro avg       0.68      0.55      0.56       441
weighted avg       0.80      0.84      0.80       441

Accuracy score: 0.8390022675736961
ROC AUC:  0.5512181195279787


Random forest model have 100% accuracy score on training dataset and also high accuracy score on test dataset if train dataset is large. If the train set is smaller, Random forest still have 100% accuracy score on training dataset but low accuracy score on test dataset.

In [21]:
# SVM
svm_clf = SVC(kernel='linear')
svm_clf.fit(X_train, y_train)

print('TRAINING RESULTS')
print('=======================')
print('Classification report')
print(classification_report(y_train,svm_clf.predict(X_train)))
print('Accuracy score:',accuracy_score(y_train, svm_clf.predict(X_train)))
print('ROC AUC: ', roc_auc_score(y_train,svm_clf.predict(X_train)))
print('\nTESTING RESULTS')
print('=======================')
print('Classification report')
print(classification_report(y_test,svm_clf.predict(X_test)))
print('Accuracy score:',accuracy_score(y_test, svm_clf.predict(X_test)))
print('ROC AUC: ', roc_auc_score(y_test,svm_clf.predict(X_test)))

TRAINING RESULTS
Classification report
              precision    recall  f1-score   support

           0       0.95      0.99      0.97       863
           1       0.94      0.72      0.81       166

    accuracy                           0.95      1029
   macro avg       0.94      0.85      0.89      1029
weighted avg       0.95      0.95      0.94      1029

Accuracy score: 0.9465500485908649
ROC AUC:  0.8537987407335018

TESTING RESULTS
Classification report
              precision    recall  f1-score   support

           0       0.89      0.93      0.91       370
           1       0.52      0.38      0.44        71

    accuracy                           0.84       441
   macro avg       0.70      0.66      0.67       441
weighted avg       0.83      0.84      0.83       441

Accuracy score: 0.8435374149659864
ROC AUC:  0.6563570612866387


SVM model have very high accuracy score on training dataset and also very high accuracy score on test dataset if train dataset is large. If the train set is smaller, SVM still have high accuracy score on training dataset and test dataset but SVM performs better with large train dataset, perhaps the best model out of 3 - compare to logistic regression and random forest.